File: Solution/1608.特殊数组的特征值/1608.特殊数组的特征值.py
'''
Descroption: LeetCode 1608. 特殊数组的特征值
Author: EmoryHuang
Date: 2021-06-24 15:50:00
解题思路:
排序，然后寻找 x，满足前一个数小于 x 后一个数大于等于 x
'''


class Solution:
    def specialArray(self, nums: List[int]) -> int:
        nums.sort()
        n = len(nums)
        if nums[0] >= n:
            return len(nums)  # n 个数大于等于 n
        for i in range(1, n + 1):
            # i 个数大于等于 i
            if nums[n - i] >= i and nums[n - i - 1] < i:
                return i
        return - 1

File: Solution/1608.特殊数组的特征值/1608.特殊数组的特征值.java
/*
 * @Descroption: LeetCode 1608. 特殊数组的特征值
 * @Author: EmoryHuang
 * @Date: 2022-09-12 11:16:51
 * @Method:
 * 排序 + 枚举 + 二分
 * 枚举x，通过排序加二分找到nums中比x大的元素个数
 */


class Solution {
    public int specialArray(int[] nums) {
        Arrays.sort(nums);
        for (int x = 0; x <= 1000; x++) {
            int l = 0, r = nums.length - 1;
            while (l < r) {
                int mid = (l + r) / 2;
                if (nums[mid] >= x)
                    r = mid;
                else
                    l = mid + 1;
            }
            if (nums[l] >= x && x == nums.length - r)
                return x;
        }
        return -1;
    }
}